samza-core/src/main/java/org/apache/samza/job/model/JobModel.java

```diff
@@ -119,6 +125,30 @@ public String getContainerToHostValue(String containerId, String key) {
     return allValues;
   }
 
+  /**
+   * Extracts the map of {@link SystemStreamPartition}s to {@link TaskName} from the underlying {@link TaskModel}s
+   *
+   * @return the extracted map
+   */
+  public Map<TaskName, Set<SystemStreamPartition>> getTaskToSystemStreamPartitions() {
+    HashMap<TaskName, Set<SystemStreamPartition>> taskToSSPs = new HashMap<>();
+    for (String containerID : containers.keySet()) {
+      ContainerModel containerModel = containers.get(containerID);
+      for (TaskName taskName : containerModel.getTasks().keySet()) {
+        TaskModel taskModel = containerModel.getTasks().get(taskName);
+        if (taskModel.getTaskMode() != TaskMode.Active) {
+          // Avoid duplicate tasks
+          continue;
+        }
+        ImmutableSet.Builder<SystemStreamPartition> setOfSSPs = ImmutableSet.builder();
+        setOfSSPs.addAll(taskToSSPs.getOrDefault(taskName, ImmutableSet.of()));
```

Comment:
Typically it is, but not always. If you look at the `SingleContainerGrouperFactory`, it maps all tasks to all containers. Even though that grouper is supposed to only be used by the `PassthroughCoordinator` which can't support startpoints anyway, there is no strict enforcement on container to task mappings.

Reply:
Thanks for clarifying.
The pattern you are using for building up the SSP sets (i.e. only using immutable sets and then creating new ones when they need to be updated) seems a little inconsistent with what I have seen more regularly in Java. Usually, the value is a mutable set, and then you just add to that set. The pattern you are using is more of a Scala pattern, since Scala better supports using immutable sets. If you stick with this pattern, could you please put a short note about why this works for what you need?

Reply:
Cool if it fits the Java pattern better, I'll switch to the mutable type. No strong requirement other than following the functional programming (Scala) pattern.

Reply:
If you wanted to, you could still wrap the end result in immutable structures. But not necessary.

Reply:
I'll leave it mutable for local consistency.